lib/streamlit/elements/slider.py

```diff
@@ -33,6 +34,83 @@
+        if ui_value is not None:
+            val = ui_value
+        else:
+            # Widget has not been used; fallback to the original value,
+            val = self.value
+
+        # The widget always returns a float array, so fix the return type if necessary
+        if self.data_type == SliderProto.INT:
+            val = [int(v) for v in val]
+        if self.data_type == SliderProto.DATETIME:
+            val = [self._micros_to_datetime(int(v)) for v in val]
+        if self.data_type == SliderProto.DATE:
+            val = [self._micros_to_datetime(int(v)).date() for v in val]
+        if self.data_type == SliderProto.TIME:
+            val = [
+                self._micros_to_datetime(int(v)).time().replace(tzinfo=self.orig_tz)
+                for v in val
+            ]
+        return val[0] if self.single_value else tuple(val)
+
+    def serialize(self, v: Any) -> List[Any]:
+        range_value = isinstance(v, (list, tuple))
+        value = list(v) if range_value else [v]
+        if self.data_type == SliderProto.DATE:
+            value = [self._datetime_to_micros(self._date_to_datetime(v)) for v in value]
+        if self.data_type == SliderProto.TIME:
+            value = [self._datetime_to_micros(self._time_to_datetime(v)) for v in value]
+        if self.data_type == SliderProto.DATETIME:
+            value = [self._datetime_to_micros(v) for v in value]
+        return value
+
+    def _time_to_datetime(self, time):
+        # Note, here we pick an arbitrary date well after Unix epoch.
+        # This prevents pre-epoch timezone issues (https://bugs.python.org/issue36759)
+        # We're dropping the date from datetime laters, anyways.
+        return datetime.combine(date(2000, 1, 1), time)
+
+    def _date_to_datetime(self, date):
+        return datetime.combine(date, time())
+
+    def _datetime_to_micros(self, dt):
+        # The frontend is not aware of timezones and only expects a UTC-based timestamp (in microseconds).
+        # Since we want to show the date/time exactly as it is in the given datetime object,
+        # we just set the tzinfo to UTC and do not do any timezone conversions.
+        # Only the backend knows about original timezone and will replace the UTC timestamp in the deserialization.
+        utc_dt = dt.replace(tzinfo=timezone.utc)
+        return self._delta_to_micros(utc_dt - UTC_EPOCH)
+
+    def _delta_to_micros(self, delta):
+        return (
+            delta.microseconds
+            + delta.seconds * SECONDS_TO_MICROS
+            + delta.days * DAYS_TO_MICROS
+        )
+
+    # Restore times/datetimes to original timezone (dates are always naive)
+    def _micros_to_datetime(self, micros):
+        utc_dt = UTC_EPOCH + timedelta(microseconds=micros)
+        # Add the original timezone. No conversion is required here,
+        # since in the serialization, we also just replace the timestamp with UTC.
+        return utc_dt.replace(tzinfo=self.orig_tz)
+
```

Comment:
It looks like these functions all still exist as closures inside the `_slider` function - are they still used there? If so, can we de-dupe them and move them up to the top of the file?